### PR TITLE
Paint 2.0

### LINKED ISSUE #27: Port MFPaint
**Summary**
_What is the feature all about? Is it a game, a utility, an enhancement, or something else?_
It's the MFPaint you know and love, on the new swadge! Plus new features and improvements.

Current roadmap as of October 3rd:

Planned Work:
- [x] Replace brush selection with tool wheel
- [x] Replace brush size with tool wheel
- [x] Replace save/load/exit menus with tool wheel
- [ ] New UI for palette editing
- [x] Revamp saving to use heatshrink compression
- [x] Revamp saving to allow arbitrary numbers of saves
- [ ] Add a total space limit for images, so we don't use _all_ the NVS space
- [x] Restructure canvas system so images don't only live in the framebuffer
- [ ] Custom canvas dimensions, or just selectable
- [x] Draw previews of shapes
- [ ] Wired sharing option
- [ ] Pre-loaded theme-related images

Stretch:
- [ ] Select/Copy/Paste
- [ ] 90-degree rotation
- [ ] Color replace
- [ ] Gallery (screensaver) mode improvements (background color, tiling, etc)
- [ ] Stamps
- [ ] Text

**Technical Spec**
_How will this be implemented?_
_What peripherals will be used, and how?_
Touch support will need to be redone, due to the new touchpad design. This will probably make it much nicer, especially if iPod-style rotation is a thing (think "tool wheel")

The menus will likely also be completely redone to make use of the new generic menus with a custom renderer. And in general, the UI will likely be almost entirely redone as the main constraint the orginal UI had was "never draw over top of the canvas, because the image doesn't really fit outside the framebuffer" which is now irrelevant thanks to the vastness and fast-enough-ness of SPIRAM.

**How to Test**
How can we verify the feature is working?

Compare with last year's version

### PR DESCRIPTION
**Pull request by dylwhich**
### Description

Basically the final version of MFPaint. Highlights:
* New touchpad support
* Named, compressed image files!
* Previews of drawing shapes
* Improved gallery
* Images live in memory, not just the framebuffer, yay!
* I deleted like half the code, so it's gotta be better!

### Test Instructions

Play around with it, make some art, do weird things! Try to break it!

### Ticket Links

Closes #27 

### Readiness Checklist
- [x] I have run `make format` to format the changes
- [x] I have compiled the firmware and the changes have no warnings
- [x] I have compiled the emulator and the changes have no warnings
- [x] I have run `make cppcheck` and checked that `cppcheck_result.txt` has no warnings for the changes
- [x] I have added doxygen comments to any code used by more than one Swadge mode. This includes `/*! \file` comments with Design Philosophy, Usage, and Example sections for new headers.
- [x] I have run `make docs` and checked that `doxy_warnings.txt` has no warnings for the new code

### CODE REVIEW

**Comment by AEFeinstein** [comment]
Is this ready for review or should I wait for the checklist?

**Comment by dylwhich** [comment]
Ready for review now! Just had to fix a couple things last-night me forgot about.

File: main/asset_loaders/common/heatshrink_encoder.c
Comment (AEFeinstein) [inline main/asset_loaders/common/heatshrink_encoder.c:1]: You madman, you added the encoder!

File: main/asset_loaders/heatshrink_helper.c [hunk 1]
```diff
@@ -81,3 +87,198 @@ uint8_t* readHeatshrinkFile(const char* fname, uint32_t* outsize, bool readToSpi
...
+        return NULL;
+    }
+
+    if (!readNamespaceNvsBlob(namespace, key, buf, &sz))
+    {
+        free(buf);
+        return NULL;
+    }
+
+    // Pick out the decompresed size and create a space for it
```
Comment (AEFeinstein) [inline main/asset_loaders/heatshrink_helper.c:113]: You should deduplicate the decompression. It seems to be shared whether it's coming from SPIFFS or NVS

File: main/display/wsg.c [hunk 2]
```diff
@@ -323,6 +324,82 @@ void drawWsgSimple(const wsg_t* wsg, int16_t xOff, int16_t yOff)
...
+/**
+ * @brief Draw a WSG to the display without flipping or rotation
+ *
+ * @param wsg  The WSG to draw to the display
+ * @param xOff The x offset to draw the WSG at
+ * @param yOff The y offset to draw the WSG at
+ * @param xScale The amount to scale the image horizontally
+ * @param yScale The amount to scale the image vertically
+ */
+void drawWsgSimpleScaled(const wsg_t* wsg, int16_t xOff, int16_t yOff, int16_t xScale, int16_t yScale)
```
Comment (AEFeinstein) [inline main/display/wsg.c:336]: Would be nice to combine rotation, reflection, and scaling one day. But not today.

File: main/utils/textEntry.h [hunk 3]
```diff
@@ -0,0 +1,143 @@
+/*! \file textEntry.h
+ *
+ * \section textEntry_design Design Philosophy
+ *
+ * The text entry utility provides a minimal interface for editing a single line of text of unlimited length.
+ * It can easily be adapted to any interface
+ *
+ * \section textEntry_usage Usage
```
Comment (AEFeinstein) [inline main/utils/textEntry.h:8]: A little sparse

**Comment by AEFeinstein** [comment]
Brush size requiring a touchpad and a button at the same time is not intuitive. It should have a second level touchpad menu like the other options

**Comment by AEFeinstein** [comment]
Also I dont think the filename is being set into `dialogUnsavedDetailStr` properly if you save as something non-default first.

File: main/asset_loaders/heatshrink_helper.c (hunk 1 above)
Comment (dylwhich) [reply main/asset_loaders/heatshrink_helper.c:113]: > You should deduplicate the decompression. It seems to be shared whether it's coming from SPIFFS or NVS

Ah yeah, I think I meant to do that at some point...

File: main/display/wsg.c (hunk 2 above)
Comment (dylwhich) [reply main/display/wsg.c:336]: Ah, yeah I was going to just add a function with a transform matrix but then i realized it was a Bit More Complicated and just did this instead.

File: main/utils/textEntry.h (hunk 3 above)
Comment (dylwhich) [reply main/utils/textEntry.h:8]: Woops. That reminds me I need to add control displays here too...

**Comment by dylwhich** [comment]
> Brush size requiring a touchpad and a button at the same time is not intuitive. It should have a second level touchpad menu like the other options

Ah yeah, this is one thing I'm hoping to be able to get to today.


> Also I dont think the filename is being set into `dialogUnsavedDetailStr` properly if you save as something non-default first.

Good catch, thanks!

**Comment by dylwhich** [comment]
Ok, fixed the one issue but looks like I'll work on the rest of it tomorrow. Should have plenty of time!

**Comment by AEFeinstein** [comment]
@dylwhich, turns out @cnlohr slipped in a completely different `textEntry.c` for flightsim before this one. I'd strongly recommend renaming yours rather than trying to integrate or anything, and we can hash it out next year.

**Comment by dylwhich** [comment]
> @dylwhich, turns out @cnlohr slipped in a completely different `textEntry.c` for flightsim before this one. I'd strongly recommend renaming yours rather than trying to integrate or anything, and we can hash it out next year.

Okay, sorted this out too. That should be everything.

**Review by AEFeinstein — APPROVED** [review]

**Pull request merged**